libraries/libfc/include/fc/network/listener.hpp

```diff
@@ -157,120 +113,146 @@ struct listener : std::enable_shared_from_this<T> {
+/// create_session whenever a socket is accepted.
+///
+/// @details
+/// This function is used for listening on TCP or Unix socket address and creating corresponding session  when the
+/// socket is accepted.
+///
+/// For TCP socket, the address format can be <hostname>:port or <ipaddress>:port where the `:port` part  is mandatory.
+/// If only the port is specified, all network interfaces are listened. The function can listen on multiple IP addresses
+/// if the specified hostname is resolved to multiple IP addresses; in other words, it can create more than one
+/// fc::listener objects. If port is not specified or none of the resolved address can be listened, an std::system_error
+/// with std::errc::bad_address error code will be thrown.
+///
+/// For Unix socket, this function will temporary change current working directory to the parent of the specified \c
+/// address (i.e. socket file path), listen on the filename component of the path, and then restore the working
+/// directory before return. This is the workaround for the socket file paths limitation which is around 100 characters.
+///
+/// The lifetime of the created listener objects is controlled by \c executor, the created objects will be destroyed
+/// when \c executor.stop() is called.
+///
+/// @note
+/// This function is not thread safe for Unix socket because it will temporarily change working directory without any
+/// lock. Any code which depends the current working directory (such as opening files with relative paths) in other
+/// threads should be protected.
+///
+/// @tparam Protocol either \c boost::asio::ip::tcp or \c boost::asio::local::stream_protocol
+/// @throws std::system_error or boost::system::system_error
+template <typename Protocol, typename CreateSession>
+void create_listener(boost::asio::io_context& executor, logger& logger, boost::posix_time::time_duration accept_timeout,
+                     const std::string& address, const std::string& extra_listening_log_info,
+                     const CreateSession& create_session) {
+   using tcp = boost::asio::ip::tcp;
+   if constexpr (std::is_same_v<Protocol, tcp>) {
+      auto [host, port] = split_host_port(address);
+      if (port.empty()) {
+         fc_elog(logger, "port is not specified for address ${addr}", ("addr", address));
+         throw std::system_error(std::make_error_code(std::errc::bad_address));
+      }
 
-         int                          listened = 0;
-         std::optional<tcp::endpoint> unspecified_ipv4_addr;
-         bool                         has_unspecified_ipv6_only = false;
+      boost::system::error_code ec;
+      tcp::resolver             resolver(executor);
+      auto                      endpoints = resolver.resolve(host, port, tcp::resolver::passive, ec);
+      if (ec) {
+         fc_elog(logger, "failed to resolve address: ${msg}", ("msg", ec.message()));
+         throw std::system_error(ec);
```

Comment:
Is there a reason to use this `if/fc_elog/throw` combo for these over a single `EOS_ASSERT` line?